pkg/controller/mattermostrestoredb/mattermostrestoredb_controller.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileMattermostRestoreDB implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileMattermostRestoreDB{}
+
+// ReconcileMattermostRestoreDB reconciles a MattermostRestoreDB object
+type ReconcileMattermostRestoreDB struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	state  mattermostv1alpha1.RestoreState
+}
+
+func (r *ReconcileMattermostRestoreDB) setRestoring() {
+	r.state = mattermostv1alpha1.Restoring
+}
+
+func (r *ReconcileMattermostRestoreDB) setFinished() {
+	r.state = mattermostv1alpha1.Finished
+}
+
+func (r *ReconcileMattermostRestoreDB) setFailed() {
+	r.state = mattermostv1alpha1.Failed
+}
+
+// Reconcile reads that state of the cluster for a MattermostRestoreDB object and makes changes based on the state read
+// and what is in the MattermostRestoreDB.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileMattermostRestoreDB) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling MattermostRestoreDB")
+	// Fetch the MattermostRestoreDB instance
+	restoreMM := &mattermostv1alpha1.MattermostRestoreDB{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, restoreMM)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			r.setRestoring()
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		r.setRestoring()
+		return reconcile.Result{}, err
+	}
+
+	// Check if this Mattermost ClusterInstallation exists
+	clusterInstallation := &mattermostv1alpha1.ClusterInstallation{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: restoreMM.Spec.MattermostClusterName, Namespace: restoreMM.Namespace}, clusterInstallation)
+	if err != nil && errors.IsNotFound(err) {
+		r.setFailed()
+		status := restoreMM.Status
+		status.State = r.state
+		//TODO: add reason and inform need to delete/apply when a clusterinstallation is ready. JIRA MM-18633
+		err = r.updateStatus(restoreMM, status, reqLogger)
+		reqLogger.Error(err, "Mattermost Installation not found. Create a ClusterInstallation first", "Namespace", restoreMM.Namespace, "ClusterInstallation.Name", restoreMM.Spec.MattermostClusterName, "RestoreDB.name", restoreMM.Name)
```

Comment:
By the time we log here, we have changed the value of `err` on line 116. We should log before setting the status and then log again after if another error occurs.